include/revng/Pipeline/InspectableLLVMContainer.h

```diff
@@ -0,0 +1,171 @@
+    getRegisteredInspectors().push_back(this);
+  }
+  virtual ~LLVMContainerInspector() = default;
+
+  virtual std::optional<Target>
+  symbolToTarget(const llvm::Function &Symbol) const = 0;
+
+  static bool hasOwner(const llvm::Function &Symbol) {
+    for (const auto &I : getRegisteredInspectors())
+      if (I->symbolToTarget(Symbol))
+        return true;
+    return false;
+  }
+
+  virtual TargetsList
+  compactTargets(const Context &Ctx, TargetsList::List &Targets) const = 0;
+
+  llvm::DenseSet<const llvm::Function *>
+  enumerateFunctions(const TargetsList &Targets,
+                     const LLVMContainer &Container) const {
+
+    llvm::DenseSet<const llvm::Function *> ToReturn;
+    for (auto &GL : Container.getModule().functions()) {
+      auto MaybeTarget = symbolToTarget(GL);
+      if (not MaybeTarget.has_value())
+        continue;
+
+      if (not Targets.contains(*MaybeTarget))
+        continue;
+
+      ToReturn.insert(&GL);
+    }
+
+    return ToReturn;
+  }
+
+  llvm::DenseSet<llvm::Function *>
+  enumerateFunctions(const TargetsList &Targets,
+                     LLVMContainer &Container) const {
+    llvm::DenseSet<llvm::Function *> ToReturn;
+    for (auto &GL : Container.getModule().functions()) {
+      auto MaybeTarget = symbolToTarget(GL);
+      if (not MaybeTarget.has_value())
+        continue;
+
+      if (not Targets.contains(*MaybeTarget))
+        continue;
+
+      ToReturn.insert(&GL);
+    }
+
+    return ToReturn;
+  }
+
+  bool remove(const Context &Ctx,
+              const TargetsList &Targets,
+              LLVMContainer &Container) const final {
+
+    const bool AllContained = enumerate(Ctx, Container).contains(Targets);
+
+    for (auto &GL : enumerateFunctions(Targets, Container))
+      GL->deleteBody();
+
+    return AllContained;
+  }
+
+  TargetsList
+  enumerate(const Context &Ctx, const LLVMContainer &Container) const final {
+    TargetsList::List L;
+    for (auto &GL : Container.getModule().functions()) {
+      auto MaybeTarget = symbolToTarget(GL);
+      if (not MaybeTarget.has_value())
+        continue;
+
+      L.push_back(std::move(*MaybeTarget));
+    }
+
+    return compactTargets(Ctx, L);
+  }
+
+  static llvm::DenseSet<const llvm::Function *>
+  enumerateNonOnwedFunctions(const LLVMContainer &Container) {
+    llvm::DenseSet<const llvm::Function *> ToReturn;
+
+    for (const auto &F : Container.getModule().functions())
+      if (not hasOwner(F))
+        ToReturn.insert(&F);
+
+    return ToReturn;
+  }
+
+  static llvm::DenseSet<const llvm::Function *>
+  enumerateAllFunctions(const TargetsList &Targets,
+                        const LLVMContainer &Container) {
+    llvm::DenseSet<const llvm::Function *> ToReturn;
+
+    for (const auto &Kind : getRegisteredInspectors()) {
+      auto Set = Kind->enumerateFunctions(Targets, Container);
+      for (const auto &E : Set)
+        ToReturn.insert(E);
+    }
+
+    return ToReturn;
+  }
```

Comment:
The `enumerateNonOnwedFunctions` and `enumerateAllFunctions` seem to refine the results of `enumerate`, but they do not.

I'd choose another prefix. What about simply `functions()` and `untrackedFunctions()`?